### DISCUSSION TITLE
OpenHasp designer

### DISCUSSION THREAD

**Discussion by mvturnho** [HASwitchPlate/openHASP#479] [Show and tell]
Hi, 

I started work on a webbased openhasp designer. It is in its early stages and build on:
- Vuejs
- Konva
- Tailwind

I have some basics working, a number of Objects work and import/export works, 
The UI is still very rough.
Let me know if other efforts have been made that are maybe better or if we can make this into something more serious.
One thing is for sure, we need a graphical editor for OpenHASP.
You can play with it here http://haspdesigner.qrisonline.nl/ , be warned it is not as userfriendly as you may want it to be, it is at most a proof of concept for the use of the choosen frameworks.

Sourcecode will be opensource like OpenHASP when the project/code is ready for this. 

![image](https://user-images.githubusercontent.com/2076073/236678280-2d75b3dd-3e56-4736-908d-13d31f166028.png)

**Comment by dgomes** [comment]
Love it!

**Reply by snelweg** [reply]
wow, seems to work for me. Thank you very much for providing this. I use it with Tasmota and HaspMota.

**Comment by adorobis** [comment]
Looks great! Some glitches still there but looks very promissing!

**Comment by jazzmonger** [comment]
Awesome! This will catapult this project forward.  

It all loads for me and I could place objects, just not edit  them.  Ill be watching this repository for sure!

**Comment by michaelsidenius** [comment]
Exactly what I was looking for! Great!

**Comment by mvturnho** [comment]
Did an update last weekend. You can change object properties with the properties item.
![image](https://github.com/HASwitchPlate/openHASP/assets/2076073/0b6d6330-d5ce-4e0d-8fb1-ff6cf1b807ee)

**Reply by jazzmonger** [reply]
Cool.  Progress!

**Comment by cherowley** [comment]
Fantastic work!

This project alone could really spread update of openhasp!

**Comment by mvturnho** [comment]
New update available. Time is limited so progress is slow.
Buttonmatrix is under development. Wrestling with how to add buttons in the correct position in the matrix.

![image](https://github.com/HASwitchPlate/openHASP/assets/2076073/6d6d8bd1-ff0e-477f-91df-441a227b23f4)

**Reply by jazzmonger** [reply]
Yeah, but there's progress!  Is there no one who can help with coding?  I'm sure all of us will be willing testers, so at least there's that!

**Comment by Digityle** [comment]
This is so awesome, been waiting for something like this! No more painful drawing everything in GUIslicer then translating manually like a caveman with a stone tablet!!

**Reply by jazzmonger** [reply]
You forgot the part about rubbing sticks together to get flame.... countless hours moving objects a few pixels at a time gets old quick...

**Reply by mvturnho** [reply]
Oh sorry stone tablets are not supported ... yet

**Comment by nagyrobi** [comment]
How does this work? Tried 2 things:
- design something new - can't seem to be able to add any objects to the page (Firefox, Chrome tested)
- opened my own manually created .jsonl - nothing happened

**Comment by mvturnho** [comment]
Works fine here. First add a page.
Then select the page to add objects. The selected page has a red border.
Add a button and move / resize it. With the properties menu yoy can chage border label etc.
Select the page again and add Object. Then select the created object to add for example a button to the object.

![image](https://github.com/HASwitchPlate/openHASP/assets/2076073/20dc28ab-53ab-4fde-b1e6-200c2ab54665)
![image](https://github.com/HASwitchPlate/openHASP/assets/2076073/7afb6b89-5f7a-4e9a-814f-30d40338277e)
![image](https://github.com/HASwitchPlate/openHASP/assets/2076073/ee6466f6-f310-42f7-aa98-1780bb7cc089)
![image](https://github.com/HASwitchPlate/openHASP/assets/2076073/8cef8ac1-defa-4859-95ab-47d3511ce212)
![image](https://github.com/HASwitchPlate/openHASP/assets/2076073/243e552e-e649-4a04-8c79-e2256e2723e7)

**Comment by mvturnho** [comment]
Also;
![image](https://github.com/HASwitchPlate/openHASP/assets/2076073/176f8ffe-43b7-473b-b8ab-1dc6a36a330e)

**Reply by nagyrobi** [reply]
Oh I see. I wasn't aware that the first button is a page.

**Comment by nagyrobi** [comment]
Any idea why can't I import my existing pages?
See [this](https://github.com/HASwitchPlate/openHASP/files/11746186/pages_halo.zip).

**Reply by mvturnho** [reply]
Ok tested this and it seems there is some more work needed to import this correctly.
first you should define the page separetly:
![image](https://github.com/HASwitchPlate/openHASP/assets/2076073/0ffef87a-dfdf-4b80-bef0-2da66afa491a)

then font sizes do not work correct yet.
Images can only have weblocations
and much more ... I gues

But after some fixes I was able to import your jsonl. (the fixes are not online yet) 
I will publish the code shortly maybe with some help, improvement goes faster,

You could try rebuild you setup with the limitations of the designer and see hoe the export shows on your plate.
You can also change the id's of your objects from the properties menu.

![image](https://github.com/HASwitchPlate/openHASP/assets/2076073/d09194a2-314c-4a5a-9aff-e0f39384de1e)

![image](https://github.com/HASwitchPlate/openHASP/assets/2076073/cefe5f9d-fb5e-4f7e-8065-7d1958c6fa8f)

**Reply by mvturnho** [reply]
And by the way, BIG thanks for testing the designer

**Reply by nagyrobi** [reply]
It looks like there are some rendering misunderstandings. For example on the page "IDŐJÁRÁS" the tabview has invisible tab labels, the designer should keep that in mind. (switching between the tabs happens by dragging).
Missing pictures should only be displayed with placeholders.
Missing fonts (rendered from TTF directly) should use some default font on the designer.

I also suggest that it should be possible to have a kind of "demo text" field for the labels, values etc. where the text itself wouldn't be exported in the final jsonl, but it would allow us to test basically how it would look like with it. Use case is for texts which are rendered dynamically by the custom component.

**Comment by cimburlen** [comment]
You finally did it!  Wow, was waiting for something similar.  Bought a Sunton 7 display a couple of months ago, for this project, but when I tried to set it up, my head exploded and I spat on it.  You lit a fire in me, thank you, I look forward to the development of your project.

**Comment by jazzmonger** [comment]
@mvturnho I created a new page today for my Alexa Timers. It all worked!  Do you have plans to allow font sizes and type definitions?

importing my existing file failed spectacularly (!), but it's quite complex.

**Reply by mvturnho** [reply]
Hi, great that you tested the designer.
There are lots of things that need to be done. Fonts is one if them. The reason why most imports fail is because  the current import expects a separate line to define a page and then the objects for that page. I have an update available that I will deploy shortly. Then I wil do some code cleanup and make it publicly available on github  so others can maybe help improve and expand this.

**Reply by jazzmonger** [reply]
Yes, each of my pages have only one "page" defined on the first line.

```
{"page":4,"comment":"---------- CAMERAS  ----------"}

{"id":1,"obj":"img","src":"L:/openhasp.png","w":240,"h":240,"x":-140,"y":40}
{"id":2,"obj":"img","src":"L:/openhasp.png","w":240,"h":240,"x":340,"y":40}

{"page":5,"comment":"----------- ALEXA ACTIVE TIMERS  ------- "}
{"id":1,"obj":"label","x":140,"y":200,"w":151,"h":40,"border_color":"#848a94","border_width":0,"radius":5,"bg_color":"#5a6173","text":"Alexa Timer #1","text_color":"white"}
Etc...
```

**Comment by cook1emr** [comment]
This looks very promising. Thanks for spending your time on it.

Will you make this selfhostable in future? 

In the meanwhile can you server this over https please?

**Reply by mvturnho** [reply]
Ah I see... It's https now. Although no data is send to the server.

**Comment by jazzmonger** [comment]
@mvturnho You should consider creating your own GitHub repository project for this designer. Having multiple discussion areas and issue trackers could really help supercharge this project. You might get additional code collaboration from others as well.

**Reply by cook1emr** [reply]
I would love to see this evolve into a selfhosted webserver. That could be connected to my various panels and provide central administration along with design. 

I feel that having a selfhosted option would be in keeping with the original project.

**Comment by htvekov** [comment]
@mvturnho 

Just had a quick glance on your designer.
Great work so far 💪😎
It's a massive task, but this will really help a lot of users (both new and experienced) in the design phase.

**Comment by oradke** [comment]
Hey that looks like a GREAT project! Would be happy to contribute.

Question: Would it be possible to copy/paste config code instead of uploading/downloading files? And see the code while working in the designer?

**Comment by mvturnho** [comment]
Hi All, it has been a while after my last update, and I have now released the sourcecode in its own repository you can find it here.
https://github.com/mvturnho/OpenHaspDesigner

I hope there are interested developers that are willing to join and help this project foreward.

**Reply by jazzmonger** [reply]
Well done, nice and complete! You might want to enable the discussions feature on your project. You can document things and interact with other devs along the way as you make progress. I used it to great effect when I designed my very complex pellet stove controller.

Too bad I'm a hack programmer, I'd love to help but it's way over my abilities.  Your tool is something that will be REALLY useful to new and existing users once it matures a bit and can handle importing our advanced configs. I always dread manually going in and making changes to my plates... it's such a huge time sink. The end result is always satisfying, but the process is brutally painful.
Jeff

**Comment by smcgann99** [comment]
Thanks for taking the time to build this, its a fantastic addition to openhasp.

Not sure if you are aware but there is a lot of interest in wall displays in the Node Red world as well. 
eg https://discourse.nodered.org/t/new-shelly-wall-display/80961/15

As the NR community are into building their own dashboards, rather than using HA etc they tend to be very knowledgeable and capable programmers.  You might want to advertise this project there as well, to drum up some support ;-)

**Comment by darkman7hr** [comment]
If I import my design pages, on export, nothing happens? Other than that, it's very nice thing you're doing!

**Comment by Maddin4000** [comment]
Hi, very Nice tool, how can i use the Homeassistant? I dont now what do i fill in the fields Event & Service?

**Reply by tofuSCHNITZEL** [reply]
this is not yet implemented I'm afraid.

**Reply by smirnowegor** [reply]
> this is not yet implemented I'm afraid.

Hi! Does it work with Home Assistant or still not?

**Reply by cook1emr** [reply]
Whilst it would be neat to be able to create the yaml using this tool. I am curious how that would be implemented? The variations of layout or   user intent feel like this would be very complex.

**Comment by nthulke** [comment]
This is amazing, does anyone else have issues setting the size of the plate? I can't seem to set it to match my plate size 480x480.

**Reply by jascdk** [reply]
have the same problem:(

**Reply by nthulke** [reply]
I played around today, set the resolution I needed then refreshed the page and it came good, worth a shot :)

**Comment by foskolo1** [comment]
Hi!

I have often used the designer, it is not perfect but very useful. But today it is no more reacheable, the web site is "dead".  Does anybody knows why?

**Reply by nthulke** [reply]
I was able to reach it using this link https://haspdesigner.qrisonline.nl/

**Reply by foskolo1** [reply]
Problem solved.
It was the browser Opera. I tried with Firefox and I can reach the site without problems.

Thanks!

**Comment by orehmann** [comment]
Hi... 

This tool is so cool. My only problem is to configure MQTT to access my HomeAssistant MQTT Mosquitto Broker. HA itself (accessing entities works fine). It seems HaspDesigner uses wss:// transport.

But my Browser always complains (Debug Mode Console)

WebSocket connection to '**wss**://192.168.1.42:8884/' failed: 

I am pretty sure it has something to do with SSL but I have valid fullchain.pem/privkey.pem files in place (ssl folder)

MQTT Explorer can successfully connect to ws://192.168.1.42:1884 but HaspDesigner seems not to support WebSockets without TLS (wss://) 

Any idea what is wrong with my config ? Or is it possible to tell HaspDesigner to not use TLS so ws:// instead of wss://

Any help would be appreciated

Regards,
Oliver

**Reply by orehmann** [reply]
> Hi...
> 
> This tool is so cool. My only problem is to configure MQTT to access my HomeAssistant MQTT Mosquitto Broker. HA itself (accessing entities works fine). It seems HaspDesigner uses wss:// transport.
> 
> But my Browser always complains (Debug Mode Console)
> 
> WebSocket connection to '**wss**://192.168.1.42:8884/' failed:
> 
> I am pretty sure it has something to do with SSL but I have valid fullchain.pem/privkey.pem files in place (ssl folder)
> 
> MQTT Explorer can successfully connect to ws://192.168.1.42:1884 but HaspDesigner seems not to support WebSockets without TLS (wss://)
> 
> Any idea what is wrong with my config ? Or is it possible to tell HaspDesigner to not use TLS so ws:// instead of wss://
> 
> Any help would be appreciated
> 
> Regards, Oliver

Can anybody help here please ?

**Comment by foskolo1** [comment]
Hi!

I'm using the designer and it works good. I miss some features ("Undo", alignement of elements...) but until now I could build good display designs. Now I have a problem with the "switch" element. I put it on the display and I can so switch (MQTT) a Tasmota plug. It works well. Unfortunately, when I switch the Tasmota plug directly with my hand or if I switch it with another software, on the display the position of the switch remains (obviously) unchanged. How can I synchronise the position of the switch on the display with the status of the Tasmota plug? 
I couldn't find a parameter or command to set the position of the switch via software.
Thany you!

**Reply by smcgann99** [reply]
"position of the switch via software" what software are you using to drive the openhasp screen ?

**Reply by adorobis** [reply]
You can do it in Home Assistant. OpenHASP anyway is fully controlled from home assistant so that's how you can achieve synchronization of any element on the screen with any entity in home assistant. Should be achievable from OpenHab I guess as well, also via MQTT I suppose.

**Reply by foskolo1** [reply]
To drive the openhasp display I use directly nodered and mqtt. It works fine. I don't need Homeassistant or other SW. With the Java-Scripts written by Chatgpt it is very simple.
Now I found a solution to my problem. I have to read the state of the switch with an mqtt input node "boiler/tele/STATE", **change "ON" in "1" and "OFF" in "0"** and send this to the switch on the display with "hasp/sunton3/command/p3b8.val". 
The change of "ON" in "1" and "OFF" in "0" seems to be important. If I send "ON" of "OFF" directly nothing happens. It is important too, to use **"boiler/tele/STATE"** to get the status of the switch. In this case the status is sent with the Teleperiod time of Tasmota. If I use **"boiler/stat/POWER"** the staus is sent only when it changes and when I e.g. reset the display the switches  remain unsynchronised until next change.

**Reply by smcgann99** [reply]
If you want the screen to display the true state immediately then you can just listen to both topics  "boiler/tele/STATE" & "boiler/stat/POWER" and update the screen. 
I also use NR and what I do is cache the state of each switch and resend them if the plate is rebooted.

**Reply by smcgann99** [reply]
You could take a look at https://discourse.nodered.org/t/using-openhasp-with-node-red/81457 for some tips.

**Comment by foskolo1** [comment]
Now I have another problem. How can I switch the display to show a specific page with an MQTT command ? Any idea?

**Reply by tofuSCHNITZEL** [reply]
So you are not using the display with home assistant at all? Because the openhasp integration in homeassistant has a service call to change the page... You could connect to the mqtt broker as well and see what message is sent when the page is changed via the service call. I don't have a setup here at the moment to check it unfortunately

**Reply by foskolo1** [reply]
I know nodered very well and with it I can do everything in a simple way. With the many addons od Nodered I can get ALL the information existing in the world and display them on the display. Nodered runs on a docker on my Synology Nas and on the the Syno runs Mosquitto too. So I don't need other hardware.

**Reply by tofuSCHNITZEL** [reply]
> Now I have another problem. How can I switch the display to show a specific page with an MQTT command ? Any idea?

I checked it with my Home assistant install:
the mqtt message looks like this:
hasp/*platename*/command/page
and the payload is the page number

**Reply by foskolo1** [reply]
Thanks! I will try it.

**Reply by foskolo1** [reply]
It works! 
Thank you!

**Reply by foskolo1** [reply]
Thank you! All problems solved.

**Comment by foskolo1** [comment]
And now another Problem. I connected an active beeper to the gpio 18 of the display. I have configured the gpio 18 as output (relay) and tried to switch the output in nodered with a MQTT output with topic hasp/plate/command/output18 and payload 1 or 0. In the Mqtt explorer I can see that the topic hasp/plate/command/output18 receives the 1 or 0 but the level of the gpio18 pin remains 0V (=no sound).  
I tried with the gpio17 too: no sound.

Any idea?
Thank you!

**Comment by jaro44** [comment]
Hello, how can I replace icons with my own colored icons.png?
openhasp.yaml:
`      - obj: "p1b2"  # ikona forecast
        properties:
          "text": >-
            {% set icons = {"cloudy":"\uE590", "fog":"\uE591", "hail":"\uE592", "lightning":"\uE593", "clear-night":"\uE594", "partlycloudy":"\uE595", "pouring":"\uE596", "rainy":"\uE597", "snowy":"\uE598", "sunny":"\uE599", "windy":"\uE59D", "windy-variant":"\uE59E", "lightning-rainy":"\uE67E", "snowy-rainy":"\uE67F", "exceptional":"\uE957"} %}
            {{ icons[states("weather.forecast_dom")] | default("\uE957") }}`
pages.jsonl:
`{"page":1,"id":2,"obj":"label","x":31,"y":32,"w":48,"h":48,"text":"\uE596", "text_font": 48}`

**Comment by TommySharpNZ** [comment]
Very cool project and I hope @mvturnho keeps at it!
At the moment I'm designing my wireframes on paper and then using photoshop to take screenshots and have to nudge stuff by pixels to get it nicely aligned. A GUI like this is very cool!

**Reply by Myumdom** [reply]
excelent!

**Comment by TesaTesa** [comment]
Hello,
absolutely super stupid question, but im really struggling here - where to set desired resolution of the "page" to fit my display? I would suppose I should find it in settings of the page compoment, but its not there...
Thanks,
Radek

**Reply by foskolo1** [reply]
"Settings" -> "UI settings" -> "Plates". In "Plate definitions" you can define the Plate (=display) you need. You can switch between them setting the display you want to use on the "Active" field.

Enzo

**Reply by TesaTesa** [reply]
Thank you ! this little button there :D ..
Again, thanks!

**Comment by ariboehme** [comment]
Thank @mvturnho, great work!!

I have a question: How do I change the screen viewing scale? I can only see the screen is either too big or too small...

<img width="782" height="302" alt="image" src="https://github.com/user-attachments/assets/e4f4f866-3d24-4cb4-90e5-238a8cdf520b" />

**Comment by Operrr** [comment]
Everything is just great, thanks for your efforts! Is there an offline version?

**Comment by TommySharpNZ** [comment]
Are you still working on this @mvturnho ???
I am making a "webHASP" project that replicates some of the great openHASP functionality but it all runs in the browser so can be run on cheap tablets and even TVs.
It's all JSON config driven but my hope is to create a drag and drop designer much like this one.

**Reply by woody4165** [reply]
> Are you still working on this @mvturnho ???
> I am making a "webHASP" project that replicates some of the great openHASP functionality but it all runs in the browser so can be run on cheap tablets and even TVs.
> It's all JSON config driven but my hope is to create a drag and drop designer much like this one.

I'm very interested, where do I find more info? 

Thanks

**Reply by TommySharpNZ** [reply]
@woody4165 I only really got stuck into it yesterday but have the following concepts and widgets working well...

- All self contained so no mqtt or addons... Just uses websockets and the Home Assistant API.
- Pages
- Rectangles
- Labels
- Buttons
- Images
- Cameras
- Arcs

Are you able to share any screenshots of your openHASP screens as I'm concious I don't want to build a solution that just targets my requirements and design style.

**Reply by woody4165** [reply]
> Are you able to share any screenshots of your openHASP screens as I'm concious I don't want to build a solution that just targets my requirements and design style.

here is my demo project that I made using openHasp with a Guiton 480480
![openhasp demo](https://github.com/user-attachments/assets/e2f8be1e-b683-45dc-a834-dbd228a78892)

It's just a demo, I'm also trying EspHome LVGL, it has more power graphic, but the code is more hard to write

**Reply by fvanroie** [reply]
Please create a new repo or thread as this is not related to the openHASP designer.

**Comment by atrent** [comment]
I've seen a downloadable version for windows and mac, is there a linux one?
